source/grid/grid_tools.cc

```diff
@@ -4352,284 +4351,102 @@ namespace GridTools
-    std::pair<typename Triangulation<dim, spacedim>::active_cell_iterator,
-              Point<dim>>
-      my_pair;
-
-    bool         found          = false;
-    unsigned int points_checked = 0;
-
-    // If a hint cell was given, use it
-    if (cell_hint.state() == IteratorState::valid)
-      {
-        try
-          {
-            my_pair = GridTools::find_active_cell_around_point(cache,
-                                                               points[0],
-                                                               cell_hint);
-            found   = true;
-          }
-        catch (const GridTools::ExcPointNotFound<dim> &)
-          {
-            missing_points_out.emplace_back(0);
-          }
-        ++points_checked;
-      }
+    // Now make a tree of indices for the points
+    const auto p_tree = pack_rtree_of_indices(points);
 
-    // The tree search returns
-    // - a bounding box covering the cell
-    // - the active cell iterator
-    std::vector<
-      std::pair<BoundingBox<spacedim>,
-                typename Triangulation<dim, spacedim>::active_cell_iterator>>
-      box_cell;
-
-    // This is used as an index for box_cell
-    int cell_candidate_idx = -1;
-    // If any of the cells in box_cell is a ghost cell,
-    // an artificial cell or at the boundary,
-    // we want to use try/catch
-    bool use_try = false;
-
-    while (!found && points_checked < np)
-      {
-        box_cell.clear();
-        b_tree.query(boost::geometry::index::intersects(points[points_checked]),
-                     std::back_inserter(box_cell));
+    // Keep track of all found points
+    std::vector<bool> found_points(points.size(), false);
 
-        // Checking box_cell result for a suitable candidate
-        cell_candidate_idx = -1;
-        for (unsigned int i = 0; i < box_cell.size(); ++i)
-          {
-            // As a candidate we don't want artificial cells
-            if (!box_cell[i].second->is_artificial())
-              cell_candidate_idx = i;
-
-            // If the cell is not locally owned or at boundary
-            // we check for exceptions
-            if (cell_candidate_idx != -1 &&
-                (!box_cell[i].second->is_locally_owned() ||
-                 box_cell[i].second->at_boundary()))
-              use_try = true;
-
-
-            if (cell_candidate_idx != -1)
-              break;
-          }
+    // Check if a point was found
+    const auto already_found = [&found_points](const unsigned int &id) {
+      AssertIndexRange(id, found_points.size());
+      return found_points[id];
+    };
 
-        // If a suitable cell was found, use it as hint
-        if (cell_candidate_idx != -1)
+    // check if the cell was already in the vector before. If so, returns its
+    // index, otherwise adds it to the list of cells, return the new index,
+    // and resize also the other two vectors. Start from last entry.
+    const auto get_local_cell_index =
```

Comment:
maybe call the function `maybe_emplace_cell_and_get_index`. The function doesn't just return an index.